lib/glob.ml

```diff
@@ -75,23 +75,30 @@ let of_string ~double_asterisk s : t =
   in
 
   let piece () =
-    if read '*'
-    then if double_asterisk && read '*'
-      then ManyMany
-      else Many
+    if read '/' then
+      if read '*' then
+        if double_asterisk && read '*' then
+          if eos () then [ManyMany; Exactly('/')]
+          else [ManyMany]
+        else [Many; Exactly('/')]
+      else [Exactly('/')]
+    else if read '*'
+      then if double_asterisk && read '*'
+        then [ManyMany]
+        else [Many]
```

Comment:
Indeed this is confusing. How about we just keep one clause for double asterisk and detect the trailing `**` case by checking that `s.[!i - 2] = /`

Reply:
I don't think this is possible, because then the `/` is already added as an `Exactly('/')` piece.
The idea is, to not add the `/` as a piece, when a `**` follows.

Reply:
Alright, but I still think what is problematic is that we are spreading the double asterisk clauses across so many branches. How about if we introduce the following primitive in addition to `read`:

```ocaml
(** [read_prefix prefix] will attempt to read [prefix] and will return [true] if it was successful.
   If it fails, it will not increment the read index. *)
val read_prefix : string -> bool
```

Then the code can be written as:

```ocaml
if double_asterisk && read_prefix "/**" then
   ..
else if consume_prefix "**" then ..
 (* etc *)

```

Do you think this will help?

Reply:
You are right, this is much cleaner now!

I decided to name the function `lookahead` instead of `read_prefix`, as the word "prefix" for me suggests looking back from the current position and not ahead.
Would be fine with renaming it though, if you don't like "lookahead".

Reply:
The problem with the name lookahead is that it is reserved for peeking into available characters but not consuming it. Your lookahead function does the consuming. `read_prefix` isn't that great either, so feel free to come up with another name.

Reply:
What do you think of `read_ahead`?